transport/oob/oob.go

```diff
@@ -0,0 +1,137 @@
+package oob
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"syscall"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+type oobWriter struct {
+	writer      io.Writer
+	oobPosition int64
+	oobByte     byte // Byte to send as OOB
+	disOOB      bool // Flag to enable disOOB mode
+}
+
+var _ io.Writer = (*oobWriter)(nil)
+
+type oobWriterReaderFrom struct {
+	*oobWriter
+	rf io.ReaderFrom
+}
+
+var _ io.ReaderFrom = (*oobWriterReaderFrom)(nil)
+
+// NewOOBWriter creates an [io.Writer] that sends an OOB byte at the specified "oobPosition".
+// If disOOB is enabled, it will apply the --disOOB strategy.
+// "oobByte" specifies the value of the byte to send out-of-band.
+func NewOOBWriter(writer io.Writer, oobPosition int64, oobByte byte, disOOB bool) io.Writer {
+	ow := &oobWriter{writer: writer, oobPosition: oobPosition, oobByte: oobByte, disOOB: disOOB}
+	if rf, ok := writer.(io.ReaderFrom); ok {
+		return &oobWriterReaderFrom{ow, rf}
+	}
+	return ow
+}
+
+func (w *oobWriterReaderFrom) ReadFrom(source io.Reader) (int64, error) {
+	reader := io.MultiReader(io.LimitReader(source, w.oobPosition), source)
+	written, err := w.rf.ReadFrom(reader)
+	w.oobPosition -= written
+	return written, err
+}
+
+func (w *oobWriter) Write(data []byte) (int, error) {
+	var written int
+	var err error
+
+	if conn, ok := w.writer.(net.Conn); ok {
+		if w.oobPosition > 0 && w.oobPosition < int64(len(data)) {
+			// Write the first part with the regular writer
+			written, err = w.writer.Write(data[:w.oobPosition])
+			if err != nil {
+				return written, err
+			}
+
+			// Send the specified OOB byte using the new sendOOBByte method
+			_ = conn
+			err = w.sendOOBByte(conn)
+			if err != nil {
+				return written, err
+			}
+			data = data[written:] // Skip the OOB byte
+		}
+	}
+	// Write the remaining data
+	n, err := w.writer.Write(data)
+	written += n
+	return written, err
+}
+
+// sendOOBByte sends the specified OOB byte over the provided connection.
+// It sets the appropriate flags based on whether disOOB mode is enabled.
+func (w *oobWriter) sendOOBByte(conn net.Conn) error {
+	// Attempt to convert to *net.TCPConn to access SyscallConn
+	tcpConn, ok := conn.(*net.TCPConn)
+	if !ok {
+		return errors.New("connection is not a TCP connection")
+	}
+
+	oobData := []byte{w.oobByte}
+	var flags int
+	if w.disOOB {
+		flags = syscall.MSG_OOB | syscall.MSG_DONTROUTE // Additional flag for disOOB mode
```

Comment:
I don't see `MSG_OOB` in the [windows documentation of syscall](https://pkg.go.dev/syscall?GOOS=windows). I believe you need https://pkg.go.dev/golang.org/x/sys/windows.

It's important for us to support Windows and have cross-platform support, so providers can specify a single strategy for all platforms.

Reply:
added